### PR TITLE
Small Image Card composable - Texts and Buttons

### PR DESCRIPTION
**Pull request by spoorthipujariadobe**
## Description

- Added required compose dependencies to messaging build.gradle file
- Added `AepTextStyle` and `AepButtonStyle` that defines client side style for text and button composables.
- Added text styles and button style for various elements in SmallImageCard to `SmallImageUIStyle`
- Added `AepTextComposable` and `AepButtonComposables` accept and use `AepTextStyle` and `AepButtonStyle` respectively
- Changed SmallImageCard to use `AepTextComposable` and `AepButtonComposables` 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepButtonComposable.kt [hunk 1]
```diff
@@ -11,66 +11,72 @@
...
- *             The text.content is mandatory for this button.
- * @param actionUrl A string representing the URL to be triggered when the button is clicked.
- *                  This is a mandatory field.
- * @param borWidth Optional. An integer representing the width of the button's border in dp.
- * @param borColor Optional. A string representing the color of the button's border in hex format (e.g., "#FFFFFF").
- *                 If not provided, the default border color is transparent.
+ * @param defaultButtonStyle The default style to be applied to the button element.
+ * @param overriddenButtonStyle The button style provided by app that overrides the default button style.
+ * @param defaultButtonTextStyle The default style to be applied to the button text.
+ * @param overriddenButtonTextStyle The style provided by app that overrides the default button text style.
```
Comment (rymorale) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepButtonComposable.kt:33]: nit: "the text style provided" to match the overriddenButtonStyle docs above

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepButtonComposable.kt [hunk 2]
```diff
@@ -11,66 +11,72 @@
...
- * @param interactId A unique identifier for the button interaction. This is a mandatory field.
- * @param text An [AepText] object that contains the text to be displayed on the button.
- *             The text.content is mandatory for this button.
- * @param actionUrl A string representing the URL to be triggered when the button is clicked.
- *                  This is a mandatory field.
- * @param borWidth Optional. An integer representing the width of the button's border in dp.
- * @param borColor Optional. A string representing the color of the button's border in hex format (e.g., "#FFFFFF").
- *                 If not provided, the default border color is transparent.
+ * @param defaultButtonStyle The default style to be applied to the button element.
+ * @param overriddenButtonStyle The button style provided by app that overrides the default button style.
```
Comment (rymorale) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepButtonComposable.kt:31]: "the button style provided by the app that..."

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepTextComposable.kt [hunk 3]
```diff
@@ -11,76 +11,74 @@
...
  * A composable function that displays a text element with customizable properties.
  *
- * @param content The text content to be displayed. This is a mandatory field.
- * @param clr Optional. A string representing the color of the text in hex format (e.g., "#FFFFFF").
- *            If not provided or invalid, defaults to black.
- * @param align Optional. A string representing the alignment of the text.
- *              Accepted values are "left", "center", and "right". Defaults to start alignment.
- * @param font Optional. An [AepFont] object containing font properties such as size, weight, and style.
+ * @param defaultStyle The default style to be applied to the text element.
+ * @param overriddenStyle The style provided by app that overrides the default text style.
```
Comment (rymorale) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepTextComposable.kt:31]: "the style provided by the app that..."

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepuitemplates/uimodels/AepButton.kt [hunk 4]
```diff
@@ -17,17 +17,17 @@ package com.adobe.marketing.mobile.aepuitemplates.uimodels
  * @property id The unique ID for the button.
  * @property actionUrl The URL to be opened when the button is clicked.
  * @property text The text to be displayed on the button, represented by an [AepText] object.
- * @property borWidth The border width of the button.
- * @property borColor The border color of the button represented as AepColor object.
- * @property bgColour The background color of the button represented as AepColor object.
+ * @property borderWidth The border width of the button.
+ * @property borderColor The border color of the button represented as AepColor object.
```
Comment (rymorale) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepuitemplates/uimodels/AepButton.kt:21]: nit: "as an AepColor object."

**Review by rymorale — APPROVED** [review]
just a few nits with the docs but LGTM overall.

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/style/SmallImageUIStyle.kt [hunk 5]
```diff
@@ -11,31 +11,51 @@
...
- * @property titleTextStyle The text style for the title, if provided.
+ * @property titleTextStyle The custom text style for the title, if provided.
  */
+
 class SmallImageUIStyle(
-    private val titleTextStyle: TextStyle? = null
+    var titleAepTextStyle: AepTextStyle? = null,
+    var bodyAepTextStyle: AepTextStyle? = null,
+    var buttonAepButtonStyle: Array<AepButtonStyle?> = arrayOfNulls(3),
+    var buttonAepTextStyle: Array<AepTextStyle?> = arrayOfNulls(3),
```
Comment (prudrabhat) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/style/SmallImageUIStyle.kt:30]: What do the list of 3 items signify here for `buttonAepButtonStyle` and `buttonAepButtonStyle` ?

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepuitemplates/uimodels/AepDismissButton.kt [hunk 6]
```diff
@@ -19,4 +19,4 @@ package com.adobe.marketing.mobile.aepuitemplates.uimodels
  * - `simple`: A simple dismiss icon.
  * - `circle`: A circular dismiss icon.
  */
-data class AepDismissButton(val style: String)
+data class AepDismissButton(val style: String? = null)
```
Comment (prudrabhat) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepuitemplates/uimodels/AepDismissButton.kt:22]: This should be made generic to and IconButton before release.
Can please we add a TODO to ensure we do not miss this.

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/SmallImageCard.kt [hunk 7]
```diff
@@ -49,49 +39,48 @@ fun SmallImageCard(
...
-                ui.getTemplate().title.let {
-                    Text(
-                        text = it?.content ?: "",
-                        style = style.getTitleTextStyle(ui.getTemplate()),
-                    )
-                }
-                Text(
-                    text = ui.getTemplate().body?.content ?: "",
-                    style = MaterialTheme.typography.body1
+            Column {
```
Comment (prudrabhat) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/SmallImageCard.kt:62]: - How are we exposing the button, and text alignments w.r.t the row and columns?
- Will you be adding Image support in a follow up PR?

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/style/SmallImageUIStyle.kt (hunk 5 above)
Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/style/SmallImageUIStyle.kt:30]: There can be a maximum of 3 buttons in the small image card. This list represents the style for each in the same order

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/SmallImageCard.kt (hunk 7 above)
Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/SmallImageCard.kt:62]: Card, Row, Column styling will be added in the next PR

File: code/messaging/build.gradle.kts [hunk 8]
```diff
@@ -30,6 +34,14 @@ aepLibrary {
         gitRepoName = "aepsdk-messaging-android"
         addCoreDependency(mavenCoreVersion)
         addEdgeDependency(mavenEdgeVersion)
+
+        addMavenDependency("org.jetbrains.kotlin", "kotlin-stdlib-jdk8", BuildConstants.Versions.KOTLIN)
+        addMavenDependency("androidx.appcompat", "appcompat", BuildConstants.Versions.ANDROIDX_APPCOMPAT)
+        addMavenDependency("androidx.compose.runtime", "runtime", BuildConstants.Versions.COMPOSE)
+        addMavenDependency("androidx.activity", "activity-compose", BuildConstants.Versions.ANDROIDX_ACTIVITY_COMPOSE)
+        addMavenDependency("androidx.lifecycle", "lifecycle-viewmodel-compose", viewModelComposeVersion)
+        addMavenDependency("androidx.lifecycle", "lifecycle-runtime-compose", runtimeComposeVersion)
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/messaging/build.gradle.kts:43]: Not needed, removed

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/style/SmallImageUIStyle.kt (hunk 5 above)
Comment (prudrabhat) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/style/SmallImageUIStyle.kt:30]: Can we make the mapping more specific and linked to the buttonId - with a `Map<String, AepButtonStyle>`, `Map<String, AepTextStyle>`

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepTextComposable.kt [hunk 9]
```diff
@@ -11,76 +11,74 @@
 
 package com.adobe.marketing.mobile.aepcomposeui.aepui.components
 
-import androidx.compose.foundation.layout.fillMaxWidth
-import androidx.compose.material.Text
+import androidx.compose.material3.Text
```
Comment (prudrabhat) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepTextComposable.kt]: I think  we can move the subpackages of  `aepui` under `aepcomposeui` and remove `aepui`.
- `com.adobe.marketing.mobile.aepcomposeui.components`
- `com.adobe.marketing.mobile.aepcomposeui.style`
- `com.adobe.marketing.mobile.aepcomposeui.util`
- `com.adobe.marketing.mobile.aepuitemplates.uimodels`

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepButtonComposable.kt [hunk 10]
```diff
@@ -11,66 +11,72 @@
...
-        elevation = elevation,
-        border = borWidth?.let { BorderStroke(it.dp, borderColor) },
-        colors = ButtonDefaults.buttonColors(backgroundColor = backgroundColor),
-        modifier = Modifier.padding(8.dp) // Modifier with padding to avoid layout issues
+        onClick = onClick,
+        modifier = overriddenButtonStyle?.modifier ?: defaultButtonStyle?.modifier ?: Modifier,
+        enabled = overriddenButtonStyle?.enabled ?: defaultButtonStyle?.enabled ?: true,
+        elevation = overriddenButtonStyle?.elevation ?: defaultButtonStyle?.elevation,
+        shape = overriddenButtonStyle?.shape ?: defaultButtonStyle?.shape
+            ?: ButtonDefaults.shape,
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepButtonComposable.kt]: Done

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepTextComposable.kt (hunk 9 above)
Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepTextComposable.kt]: Done

Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/aepui/components/AepTextComposable.kt]: Also moved `uimodels` inside `aepcomposeui` as discussed in the previous PR discussion
https://github.com/adobe/aepsdk-messaging-android/pull/281#pullrequestreview-2375729923
We can rename the packages and classes more succinctly now if needed

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/components/AepTextComposable.kt [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        else -> null
+    }
+    val fontStyle = if (font?.style?.contains("italic") == true) {
+        FontStyle.Italic
+    } else {
+        null
+    }
+
+    // Merge all text styles together
+    val mergedAepTextStyle =
```
Comment (prudrabhat) [inline code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/components/AepTextComposable.kt:73]: Would it be possible to have the composable provide the `default` and `preferred` style and do 
```
val mergedTextStyle = defaultStyle.merge(overriddenStyle)
```
and the `merge` can internally merge only once like
```
fun AepTextStyle.merge(other) {
    val textStyle = TextStyle(
        color = this.textStyle?.color ?: other.textStyle?.color ?: fallbackTextStyle.color,
        fontSize = this.textStyle?.fontSize ?: other.textStyle?.fontSize ?: fallbackTextStyle.fontSize,
        fontWeight = this.textStyle?.fontWeight ?: other.textStyle?.fontWeight ?: fallbackTextStyle.fontWeight,
        fontStyle = this.textStyle?.fontStyle ?: other.textStyle?.fontStyle ?: fallbackTextStyle.fontStyle,
    )

   return AepTextStyle(textStyle = textStyle....)
  
}
```

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/style/AepButtonStyle.kt [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+ * Returns a new [AepButtonStyle] that is a combination of this style and the other style.
+ *
+ * other AepButtonStyle's null or inherit properties are replaced with the non-null properties of this text style.
+ * Another way to think of it is that the "missing" properties of the other style are filled by the properties of this style.
+ * If this AepButtonStyle is null, returns the other AepButtonStyle.
+ * If the given AepButtonStyle is null, returns this AepButtonStyle.
+ *
+ * @param other The AepButtonStyle to merge with this AepButtonStyle
+ */
+@Composable
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/style/AepButtonStyle.kt:53]: Removed

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/style/AepTextStyle.kt [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+ * Returns a new [AepTextStyle] that is a combination of this style and the other style.
+ *
+ * other AepTextStyle's null or inherit properties are replaced with the non-null properties of this text style.
+ * Another way to think of it is that the "missing" properties of the other style are filled by the properties of this style.
+ * If this AepTextStyle is null, returns the other AepTextStyle.
+ * If the given AepTextStyle is null, returns this AepTextStyle.
+ *
+ * @param other The AepTextStyle to merge with this AepTextStyle
+ */
+@Composable
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/style/AepTextStyle.kt:48]: Removed

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/components/AepTextComposable.kt (hunk 11 above)
Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/components/AepTextComposable.kt]: Moved merge to `AepTextStyle` and `AepButtonStyle` class as discussed

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/style/SmallImageUIStyle.kt [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+class SmallImageUIStyle(
+    var titleAepTextStyle: AepTextStyle? = null,
+    var bodyAepTextStyle: AepTextStyle? = null,
+    var buttonAepButtonStyle: Array<AepButtonStyle?> = arrayOfNulls(3),
+    var buttonAepTextStyle: Array<AepTextStyle?> = arrayOfNulls(3),
+) {
+    val defaultTitleTextStyle: AepTextStyle
+        @Composable
+        @ReadOnlyComposable
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/style/SmallImageUIStyle.kt:34]: Removed

File: code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/style/AepButtonStyle.kt [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        /**
+         * Returns a new [AepButtonStyle] that is a combination of default style and overridden style.
+         * If the same property is present in all the styles, the property from the overridden style is used.
+         * If a property is not present in the overridden style, the property from the default style is used.
+         *
+         * @param defaultStyle The default [AepButtonStyle] to be applied to the button element.
+         * @param overridingStyle The [AepButtonStyle] provided by the app that overrides the default button style.
+         */
+        internal fun merge(
+            defaultStyle: AepButtonStyle? = null,
```
[the remark this answers is not in the document]
Comment (spoorthipujariadobe) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/style/AepButtonStyle.kt:53]: I provided the option to have a defaultStyle if we need to supply one in the future

[the remark this answers is not in the document]
Comment (prudrabhat) [reply code/messaging/src/main/java/com/adobe/marketing/mobile/aepcomposeui/style/AepButtonStyle.kt:53]: I see. Can we add a note explaining why the default style exists in that case.

**Review by prudrabhat — APPROVED** [review]

**Pull request merged**